include/envoy/redis/codec.h

```diff
@@ -0,0 +1,116 @@
+#pragma once
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/common/exception.h"
+
+namespace Redis {
+
+/**
+ * All RESP types as defined here: https://redis.io/topics/protocol
+ */
+enum class RespType { Null, SimpleString, BulkString, Integer, Error, Array };
+
+/**
+ * A variant implementation of a RESP value optimized for peformance. A C++11 union is used for
+ * the underlying type so that no unnecessary allocations/constructions are needed.
+ */
+class RespValue {
```

Comment:
General question about how we want to structure the code base: Is the intent for the headers under include/envoy/ to be interfaces?